### ISSUE TITLE
RFE: show when  HibernateLocation is present in systemd-boot

### ISSUE THREAD

**Issue by keszybz** [systemd/systemd#29037]
### Component

systemd-boot

### Is your feature request related to a problem? Please describe

If the machine was hibernated, the user most likely wants to resume. Show info in the boot screen that hibernation data is present. Maybe also highlight the entry or entries that will lead to a resume?

**Comment by medhefgo** [comment]
I'd go a step further: Set the loader one shot variable to the current boot entry before hibernating. Then, when both vars are set, hide any other boot entries.

This should prevent people from shooting themselves in the foot by booting the wrong kernel, or another distro or windows. They may even think it's fine to boot other installations, but in the general case it's not. It could've been mounted at hibernation time and afaik the kernel guarantees preciously little when it comes to the filesystem states during hibernation.

**Comment by medhefgo** [comment]
Also, we might as well skip timeout and menu in that case as there is no reasonable other choice to boot from and it's 99% what the user wants after having hibernated. In case of failure, the oneshot variable would've been deleted and a regular boot menu would be shown next time.

**Comment by poettering** [comment]
Yeah, I think it makes a ton of sense to shortcut a few things if the EFI var is set. In fact we probably should store info about the UKI used to boo in the EFI var, and then directly boot into that from sd-boot, instead of trying to find the entry again in the list.

**Comment by poettering** [comment]
hmm, the hibernate location efi var is json formatted, so we have three options:

1. add a full json parser to sd-boot. yuck
2. add a half-arsed json parser to sd-boot that just finds the three props we need. also yuck
3. we stop doing json. also yuck.

**Comment by poettering** [comment]
Hmm, 4th option:

distill the few things sd-boot actually cares about into a secondary efi var, which we alway set in parallel. it would only really contain the UKI name, and part uuid it is found on (or so), and nothing else.

probably the nicest option

**Comment by poettering** [comment]
ah, just referencing the UKI is probably not sufficient, we must reference the boot entry item id, because for type1 entries we need to find the .conf entry really to get all the info we need.

**Comment by poettering** [comment]
So maybe, we should on hibernation just copied the value of the existing "LoaderEntrySelected" efi var that sd-boot already sets into "LoaderEntryHibernate" or so. This should be enough for sd-boot to find selected entry again, and boot it.

**Comment by medhefgo** [comment]
> So maybe, we should on hibernation just copy the value of the existing "LoaderEntrySelected" efi var that sd-boot already sets into a new variable "LoaderEntryHibernate" or so. This should be enough for sd-boot to find selected entry again, and boot it.

I'm saying there is no need for a new variable: `LoaderEntryOneShot` already exists and effectively serves the same purpose. All you need is detect the presence of both and skip any menu/timeout (and maybe even ignore `LoaderConfigTimeoutOneShot`). The hibernation logic could just save/restore any existing value if it's already set prior to hibernation.
This would allow any boot loader that already adheres to the boot loader interface to benefit from this.

**Comment by poettering** [comment]
> I'm saying there is no need for a new variable: `LoaderEntryOneShot` already exists and effectively serves the same purpose. All you need is detect the presence of both and skip any menu/timeout (and maybe even ignore `LoaderConfigTimeoutOneShot`). The hibernation logic could just save/restore any existing value if it's already set prior to hibernation. This would allow any boot loader that already adheres to the boot loader interface to benefit from this.

I think there's value in somehow visually indicating to the user that we are in a hibrenation cycle if they enter the boot menu somehow. hence I think we should have a mechanism to clearly identify that case.

I am a bit concerned that repurposing LoaderEntryOneShot or this goal here would mean we'd take it away from what users usually want to do with it? yes, we could save/restore the value, but that sounds fragile and cumbersome. It appears to me the easier and safer way out is to just add a new efi var, and then just add it to the spec, so that others can implement this too, if they want?

**Comment by YHNdnzj** [comment]
Hmm, adding a new variable also allows people to opt out from the feature more easily. I presume there will be people that dislike the behavior of hiding other entries...

**Comment by medhefgo** [comment]
> Hmm, adding a new variable also allows people to opt out from the feature more easily. I presume there will be people that dislike the behavior of hiding other entries...

No, that's not a good reason. Those people *really* don't know what they're doing. They *think* they do, but they really don't. Hibernation is already fiddly enough as it is, let's not allow them a means to shoot themselves in the foot and lose 5h of work because they decided to take a detour through the wrong boot entry.

The variable would be deleted as soon as it's seen by sd-boot, so next time, they get a boot menu as expected.

**Comment by mskiptr** [comment]
Tbf, not filtering out Windows might be relatively safe. Its drivers don't touch Linux filesystems while `ntfs-3g` does detect if the fs is dirty | hibernated and if so refuses to mount it. And hibernating one system to boot the other is definitely something people do when it comes to this pair.

I'm not sure how WinBtrfs or FAT32 play into this.

**Comment by YHNdnzj** [comment]
> Tbf, not filtering out Windows might be relatively safe. Its drivers don't touch Linux filesystems while `ntfs-3g` does detect if the fs is dirty | hibernated and if so refuses to mount it. And hibernating one system to boot the other is definitely something people do when it comes to this pair.
> 
> I'm not sure how WinBtrfs or FAT32 play into this.

Not really. A couple of days ago I just encountered someone who broke their btrfs by booting into Windows (with WinBtrfs installed) after hibernating on Linux.
